data-prepper-plugins/aggregate-processor/src/main/java/com/amazon/dataprepper/plugins/processor/aggregate/AggregateActionSynchronizer.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package com.amazon.dataprepper.plugins.processor.aggregate;
+
+import com.amazon.dataprepper.model.event.Event;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Optional;
+import java.util.concurrent.locks.Lock;
+
+class AggregateActionSynchronizer {
+    private final AggregateAction aggregateAction;
+    private final AggregateGroupManager aggregateGroupManager;
+
+    private static final Logger LOG = LoggerFactory.getLogger(AggregateActionSynchronizer.class);
+
+    private AggregateActionSynchronizer(final AggregateAction aggregateAction, final AggregateGroupManager aggregateGroupManager) {
+        this.aggregateAction = aggregateAction;
+        this.aggregateGroupManager = aggregateGroupManager;
+    }
+
+    Optional<Event> concludeGroup(final AggregateIdentificationKeysHasher.IdentificationHash hash, final AggregateGroup aggregateGroup) {
+        final Lock concludeGroupLock = aggregateGroup.getConcludeGroupLock();
+        final Lock handleEventForGroupLock = aggregateGroup.getHandleEventForGroupLock();
+
+        Optional<Event> concludeGroupEvent = Optional.empty();
+        if (concludeGroupLock.tryLock()) {
+            try {
+                handleEventForGroupLock.lock();
+                concludeGroupEvent = aggregateAction.concludeGroup(aggregateGroup);
+                aggregateGroupManager.removeGroupWithHash(hash, aggregateGroup);
+                aggregateGroup.resetGroupStart();
```

Comment:
Should these two functions be combined? This would fully encapsulate the logic. Perhaps `aggregateGroup.closeGroup()`? 

Also, there could be value in having `removeGroupWithHash` call this function. But, that is probably not as important.

Reply:
I like the suggestion. Perhaps something like the code below for to encapsulate the closing of a group to `AggregateGroupManager`, so that `removeGroupWithHash` isn't secretly doing extra things, and `closeGroup` isn't a duplicated function name for the `AggregateGroup`. This would also enforce the idea that the `aggregateGroupManager` is the one who decides when a group is closed.

```
void closeGroup(IdentifcationHash hash, AggregateGroup group) {
       allGroups.remove(hash, group);
       group.resetGroupStart();
       group.clearGroupState();
}
```

Reply:
Yes, that is what I was thinking.

> (only thing I don't love is that there are two functions named closeGroup

Yea, I'm with you there. Maybe there is a better name for the second one.